src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -114,29 +133,62 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
       const next = completeWork(current, workInProgress);
+
+      // The work is now done. We don't need this anymore. This flags
+      // to the system not to redo any work here.
+      workInProgress.pendingProps = null;
+
+      // TODO: Stop using the parent for this purpose. I think this will break
+      // down in edge cases because when nodes are reused during bailouts, we
+      // don't know which of two parents was used. Instead we should maintain
+      // a temporary manual stack.
+      // $FlowFixMe: This downcast is not safe. It is intentionally an error.
+      const parent = workInProgress.parent;
+
+      // Ensure that remaining work priority bubbles up.
+      if (parent && workInProgress.pendingWorkPriority !== NoWork &&
+          (parent.pendingWorkPriority === NoWork ||
+          parent.pendingWorkPriority > workInProgress.pendingWorkPriority)) {
+        parent.pendingWorkPriority = workInProgress.pendingWorkPriority;
+      }
+
       if (next) {
         // If completing this work spawned new work, do that next.
         return next;
       } else if (workInProgress.sibling) {
         // If there is more work to do in this parent, do that next.
         return workInProgress.sibling;
-      } else if (workInProgress.parent) {
+      } else if (parent) {
         // If there's no more work in this parent. Complete the parent.
-        // TODO: Stop using the parent for this purpose. I think this will break
-        // down in edge cases because when nodes are reused during bailouts, we
-        // don't know which of two parents was used. Instead we should maintain
-        // a temporary manual stack.
-        // $FlowFixMe: This downcast is not safe. It is intentionally an error.
-        workInProgress = workInProgress.parent;
+        workInProgress = parent;
       } else {
         // If we're at the root, there's no more work to do.
-        currentRootsWithPendingWork = null;
-        return null;
+        console.log('completed flush with remaining work at priority', workInProgress.pendingWorkPriority);
+        if (workInProgress.pendingWorkPriority !== NoWork) {
+          // TODO: This removes all but one node. Broken.
+          currentRootsWithPendingWork = workInProgress;
+          const nextWork = findNextUnitOfWork();
+          if (!nextWork) {
+            // Something went wrong and there wasn't actually any more work.
+            // TODO: This currently fails because of the parent/root hacks.
+            // throw new Error('Should never finish with a priority level unless there is work.');
+            currentRootsWithPendingWork = null;
+            return null;
+          }
+          return nextWork;
+        } else {
+          currentRootsWithPendingWork = null;
+          return null;
+        }
       }
     }
   }
 
   function performUnitOfWork(workInProgress : Fiber) : ?Fiber {
+    // Ignore work if there is nothing to do.
+    if (workInProgress.pendingProps === null) {
```

Comment:
I was looking and performUnitOfWork is only called with nextUnitOfWork, which always comes from findNextUnitOfWork which only returns things with pendingProps.

But I guess it can be null if the reconciliation is interrupted and we perform a hi-pri update while nextUnitOfWork is set, clearing out its pending props? Let me know if I'm misunderstanding something. I don't see how traversing upwards leads to this case.